src/coreclr/tools/aot/Mono.Linker.Tests.Cases/DataFlow/MethodByRefReturnDataFlow.cs

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+using Mono.Linker.Tests.Cases.Expectations.Assertions;
+using Mono.Linker.Tests.Cases.Expectations.Helpers;
+using Mono.Linker.Tests.Cases.Expectations.Metadata;
+
+namespace Mono.Linker.Tests.Cases.DataFlow
+{
+	[SkipKeptItemsValidation]
+	[ExpectedNoWarnings]
+	public class MethodByRefReturnDataFlow
+	{
+		public static void Main ()
+		{
+			ReturnAnnotatedTypeReferenceAsUnannotated ();
+			AssignToAnnotatedTypeReference ();
+			AssignDirectlyToAnnotatedTypeReference ();
+			AssignToCapturedAnnotatedTypeReference ();
+			AssignToAnnotatedTypeReferenceWithRequirements ();
+		}
+
+		[DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.PublicMethods)]
+		static Type _annotatedField;
+
+		// This should warn, as assiging to the return ref Type will assign value to the annotated field
+		// but the annotation is not propagated
+		// https://github.com/dotnet/linker/issues/2158
+		static ref Type ReturnAnnotatedTypeReferenceAsUnannotated () { return ref _annotatedField; }
+
+		[return: DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.PublicMethods)]
+		static ref Type ReturnAnnotatedTypeReferenceAsAnnotated () { return ref _annotatedField; }
+
+		[return: DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.PublicMethods)]
+		static ref Type ReturnAnnotatedTypeWithRequirements ([DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.All)] Type t) => ref _annotatedField;
+
+		// Correct behavior in the linker, but needs to be added in analyzer
+		// Bug link: https://github.com/dotnet/linker/issues/2158
+		[ExpectedWarning ("IL2026", "Message for --TestType.Requires--", ProducedBy = ProducedBy.Trimmer)]
```

Comment:
Should we have a similar issue in NativeAOT?

Reply:
I will look into this - I should understand why it's not producing the warning (linker/AOT should behave the same in this case as it's basically the same code).

Reply:
This has been solved in later commits - it's now enabled for NativeAOT as well.